vpr/src/place/analytic_placer.h

```diff
@@ -0,0 +1,249 @@
+     */
+    void ap_place();
+
+  private:
+    // for CutSpreader to access legal_pos, block_locs, etc.
+    friend class CutSpreader;
+
+    // AP parameters that can influence it's behavior
+    struct AnalyticPlacerCfg {
+        float alpha;                        // anchoring strength of pseudo-connections
+        float beta;                         // over-utilization factor
+        int criticalityExponent;            // not currently used, @see build_equations()
+        int timingWeight;                   // not currently used, @see build_equations()
+        float solverTolerance;              // parameter of the solver
+        int spread_scale_x, spread_scale_y; // see CutSpreader::expand_regions()
+    };
+
+    AnalyticPlacerCfg tmpCfg; // TODO: PlacerHeapCfg should be externally configured & supplied
+
+    // @see initial_placement.cpp load_legal_locations()
+    std::vector<std::vector<int>> num_legal_pos;
+    std::vector<std::vector<std::vector<t_pl_loc>>> legal_pos;
+
+    // marker for blks not solved in current iteration
+    int32_t dont_solve = std::numeric_limits<int32_t>::max();
+
+    // row number in the system of linear equations for each block
+    std::unordered_map<ClusterBlockId, int32_t> blk_row_num;
+
+    // Encapsulates 3 types of locations for each logic block
+    struct BlockLocation {
+        t_pl_loc loc;       // real, up-to-date location of the logic block
+        t_pl_loc legal_loc; // legalized location, used to create psudo connections
+        double rawx, rawy;  // raw location storing float result from matrix solver
+    };
+
+    // Lookup from blockID to block location
+    std::unordered_map<ClusterBlockId, BlockLocation> blk_locs;
+
+    // The set of blks actually placed.
+    // Excludes children of macros.
+    std::vector<ClusterBlockId> place_blks;
+
+    // blocks being solved in the current equation
+    // subset of all blocks being placed
+    std::vector<ClusterBlockId> solve_blks;
+
+    /*
+     * number of blocks on each tile
+     * used to print a simple placement graph in print_place
+     */
+    std::vector<std::vector<int>> overlap;
+
+    /*
+     * Prints the location of each block, and a simple drawing of FPGA fabric, showing num of blocks on each tile
+     * Very useful for debugging
+     * See implementation for usage
+     */
+    void print_place(const char* place_file);
+
+    //build fast lookup of tiles/subtiles by tile, x, y, subtiles
+    void build_fast_tiles();
+
+    // build num_legal_pos, legal_pos like in initial_placement.cpp
+    void build_legal_locations();
+
+    // build blk_locs based on initial placement;
+    // build and place_blks for blocks that needs to be placed;
+    void init();
+
+    wirelen_t get_net_hpwl(ClusterNetId net_id);
+
+    wirelen_t total_hpwl();
+
+    // Setup the blocks of type blkTypes to be solved
+    // Returns the number of rows (number of blks to solve)
```

Comment:
Give an example; e.g. the type might be logic_blocks and this routine would ... and then return the number of rows (blocks to solve) which would be the number of logic blocks (or maybe it is the number of logic blocks that are free variables in the matrix?).